internal/api/ui/login/mfa_verify_otp_handler.go

```diff
@@ -0,0 +1,106 @@
+}
+
+type mfaOTPFormData struct {
+	Resend           bool           `schema:"resend"`
+	Code             string         `schema:"code"`
+	SelectedProvider domain.MFAType `schema:"selectedProvider"`
+	Provider         domain.MFAType `schema:"provider"`
+}
+
+func OTPLink(origin, authRequestID, code string, provider domain.MFAType) string {
+	return fmt.Sprintf("%s%s?%s=%s&%s=%s&%s=%d", externalLink(origin), EndpointMFAOTPVerify, QueryAuthRequestID, authRequestID, queryCode, code, querySelectedProvider, provider)
+}
+
+// renderOTPVerification renders the OTP verification for SMS and Email based on the passed MFAType.
+// It will send a new code to either phone or email first.
+func (l *Login) handleOTPVerification(w http.ResponseWriter, r *http.Request, authReq *domain.AuthRequest, providers []domain.MFAType, selectedProvider domain.MFAType, err error) {
+	if err != nil {
+		l.renderOTPVerification(w, r, authReq, providers, selectedProvider, err)
+		return
+	}
+	userAgentID, _ := http_mw.UserAgentIDFromCtx(r.Context())
+	sendCode := l.authRepo.SendMFAOTPSMS
+	if selectedProvider == domain.MFATypeOTPEmail {
+		sendCode = l.authRepo.SendMFAOTPEmail
+	}
+	err = sendCode(setContext(r.Context(), authReq.UserOrgID), authReq.UserID, authReq.UserOrgID, authReq.ID, userAgentID)
+	l.renderOTPVerification(w, r, authReq, providers, selectedProvider, err)
+}
+
+func (l *Login) renderOTPVerification(w http.ResponseWriter, r *http.Request, authReq *domain.AuthRequest, providers []domain.MFAType, selectedProvider domain.MFAType, err error) {
+	var errID, errMessage string
+	if err != nil {
+		errID, errMessage = l.getErrorMessage(r, err)
+	}
+	data := &mfaOTPData{
+		userData:         l.getUserData(r, authReq, "VerifyMFAU2F.Title", "VerifyMFAU2F.Description", errID, errMessage),
+		MFAProviders:     removeSelectedProviderFromList(providers, selectedProvider),
+		SelectedProvider: selectedProvider,
+	}
+	l.renderer.RenderTemplate(w, r, l.getTranslator(r.Context(), authReq), l.renderer.Templates[tmplOTPVerification], data, nil)
+}
+
+// handleOTPVerificationCheck handles form submissions of the OTP verification.
+// On successful code verification, the check will be added to the auth request.
+// A user is also able to request a code resend or choose another provider.
+func (l *Login) handleOTPVerificationCheck(w http.ResponseWriter, r *http.Request) {
+	formData := new(mfaOTPFormData)
+	authReq, err := l.getAuthRequestAndParseData(r, formData)
+	if err != nil {
+		l.renderError(w, r, authReq, err)
+		return
+	}
+	step, ok := authReq.PossibleSteps[0].(*domain.MFAVerificationStep)
+	if !ok {
+		l.renderError(w, r, authReq, err)
+		return
+	}
+	if formData.Resend {
+		l.handleOTPVerification(w, r, authReq, step.MFAProviders, formData.SelectedProvider, nil)
+		return
+	}
+	if formData.Code == "" {
+		l.renderMFAVerifySelected(w, r, authReq, step, formData.Provider, nil)
+		return
+	}
+	userAgentID, _ := http_mw.UserAgentIDFromCtx(r.Context())
```

Comment:
No reaction on !ok?

Reply:
the boolean there does not make much sense as the user agent is set in the interceptor, so there's always one (even if it changed)